src/scripts/gap_detector.py

```diff
@@ -0,0 +1,147 @@
+"""
+Gap detection script for finding missing transaction hashes of settlements.
+Uses a form of binary search to minimize/reduce API requests.
+"""
+import argparse
+import os
+
+import pandas as pd
+from pandas import DataFrame
+from dune_client.client import DuneClient
+from dune_client.query import Query
+from dune_client.types import QueryParameter
+from dotenv import load_dotenv
+
+from src.pg_client import MultiInstanceDBFetcher
+
+DB_COUNT_QUERY = (
+    "select count(*) as batches from settlements "
+    "where block_number between {{start}} and {{end}};"
+)
+DB_HASH_QUERY = (
+    "select concat('0x', encode(tx_hash, 'hex')) as tx_hash from settlements "
+    "where block_number between {{start}} and {{end}};"
+)
+
+DUNE_COUNT_QUERY_ID = 2481826
+DUNE_HASH_QUERY_ID = 2532671
+
+JANUARY_2023 = 16308190
+
+
+class GapDetector:
+    """Encapsulates all attributes and functionality of gap detection"""
+
+    def __init__(self, dune: DuneClient, database: MultiInstanceDBFetcher):
+        self.dune = dune
+        self.database = database
+
+    def dune_df(self, query_id: int, start: int, end: int) -> DataFrame:
+        """Executes and fetches dataframe from Dune `query_id`"""
+        return pd.read_csv(
+            self.dune.refresh_csv(
+                Query(
+                    query_id=query_id,
+                    params=[
+                        QueryParameter.number_type("start", start),
+                        QueryParameter.number_type("end", end),
+                    ],
+                )
+            ).data
+        )
+
+    def db_df(self, query_str: str, start: int, end: int) -> DataFrame:
+        """Executes and fetches dataframe for `query_str` from Database"""
+        return self.database.exec_query(
+            query_str.replace("{{start}}", str(start)).replace("{{end}}", str(end)),
+            self.database.connections[0],
+        )
+
+    def tx_list_diff(
+        self,
+        start: int,
+        end: int,
+    ) -> set[str]:
+        """Gets the list of differing transaction hashes between Dune and Database"""
+        dune_hashes = set(self.dune_df(DUNE_HASH_QUERY_ID, start, end)["tx_hash"])
+        db_hashes = set(self.db_df(DB_HASH_QUERY, start, end)["tx_hash"])
+        # TODO - usually its the database missing something (i.e. we assume Dune contains all).
```

Comment:
Nit: In my opinion, ideally this would reference an existing issue that we could work on in the future.

Reply:
done.